// @reactivex/rxjs/dist/cjs/operators/scan.js
var __extends = (this && this.__extends) || function (d, b) {
    for (var p in b) if (b.hasOwnProperty(p)) d[p] = b[p];
    function __() { this.constructor = d; }
    d.prototype = b === null ? Object.create(b) : (__.prototype = b.prototype, new __());
};
var Subscriber_1 = require('../Subscriber');
var tryCatch_1 = require('../util/tryCatch');
var errorObject_1 = require('../util/errorObject');
function scan(project, acc) {
    return this.lift(new ScanOperator(project, acc));
}
exports.scan = scan;
var ScanOperator = (function () {
    function ScanOperator(project, acc) {
        this.acc = acc;
        this.project = project;
    }
    ScanOperator.prototype.call = function (subscriber) {
        return new ScanSubscriber(subscriber, this.project, this.acc);
    };
    return ScanOperator;
})();
var ScanSubscriber = (function (_super) {
    __extends(ScanSubscriber, _super);
    function ScanSubscriber(destination, project, acc) {
        _super.call(this, destination);
        this.accumulatorSet = false;
        this.acc = acc;
        this.project = project;
        this.accumulatorSet = typeof acc !== 'undefined';
    }
    Object.defineProperty(ScanSubscriber.prototype, "acc", {
        get: function () {
            return this._acc;
        },
        set: function (value) {
            this.accumulatorSet = true;
            this._acc = value;
        },
        enumerable: true,
        configurable: true
    });
    ScanSubscriber.prototype._next = function (x) {
        if (!this.accumulatorSet) {
            this.acc = x;
            this.destination.next(x);
        }
        else {
            var result = tryCatch_1.tryCatch(this.project).call(this, this.acc, x);
            if (result === errorObject_1.errorObject) {
                this.destination.error(errorObject_1.errorObject.e);
            }
            else {
                this.acc = result;
                this.destination.next(this.acc);
            }
        }
    };
    return ScanSubscriber;
})(Subscriber_1.Subscriber);
