// @reactivex/rxjs/dist/cjs/Rx.js
var Observable_1 = require('./Observable');
exports.Observable = Observable_1.Observable;
// operators
/* tslint:disable:no-use-before-declare */
var combineLatest_static_1 = require('./operators/combineLatest-static');
Observable_1.Observable.combineLatest = combineLatest_static_1.combineLatest;
var concat_static_1 = require('./operators/concat-static');
Observable_1.Observable.concat = concat_static_1.concat;
var merge_static_1 = require('./operators/merge-static');
Observable_1.Observable.merge = merge_static_1.merge;
/* tslint:enable:no-use-before-declare */
var DeferObservable_1 = require('./observables/DeferObservable');
Observable_1.Observable.defer = DeferObservable_1.DeferObservable.create;
var EmptyObservable_1 = require('./observables/EmptyObservable');
Observable_1.Observable.empty = EmptyObservable_1.EmptyObservable.create;
var ForkJoinObservable_1 = require('./observables/ForkJoinObservable');
Observable_1.Observable.forkJoin = ForkJoinObservable_1.ForkJoinObservable.create;
var FromObservable_1 = require('./observables/FromObservable');
Observable_1.Observable.from = FromObservable_1.FromObservable.create;
var ArrayObservable_1 = require('./observables/ArrayObservable');
Observable_1.Observable.fromArray = ArrayObservable_1.ArrayObservable.create;
var FromEventObservable_1 = require('./observables/FromEventObservable');
Observable_1.Observable.fromEvent = FromEventObservable_1.FromEventObservable.create;
var FromEventPatternObservable_1 = require('./observables/FromEventPatternObservable');
Observable_1.Observable.fromEventPattern = FromEventPatternObservable_1.FromEventPatternObservable.create;
var PromiseObservable_1 = require('./observables/PromiseObservable');
Observable_1.Observable.fromPromise = PromiseObservable_1.PromiseObservable.create;
var IntervalObservable_1 = require('./observables/IntervalObservable');
Observable_1.Observable.interval = IntervalObservable_1.IntervalObservable.create;
var InfiniteObservable_1 = require('./observables/InfiniteObservable');
Observable_1.Observable.never = InfiniteObservable_1.InfiniteObservable.create;
Observable_1.Observable.of = ArrayObservable_1.ArrayObservable.of;
var RangeObservable_1 = require('./observables/RangeObservable');
Observable_1.Observable.range = RangeObservable_1.RangeObservable.create;
var ErrorObservable_1 = require('./observables/ErrorObservable');
Observable_1.Observable.throw = ErrorObservable_1.ErrorObservable.create;
var TimerObservable_1 = require('./observables/TimerObservable');
Observable_1.Observable.timer = TimerObservable_1.TimerObservable.create;
var zip_static_1 = require('./operators/zip-static');
Observable_1.Observable.zip = zip_static_1.zip;
var observableProto = Observable_1.Observable.prototype;
var buffer_1 = require('./operators/buffer');
observableProto.buffer = buffer_1.buffer;
var bufferCount_1 = require('./operators/bufferCount');
observableProto.bufferCount = bufferCount_1.bufferCount;
var bufferTime_1 = require('./operators/bufferTime');
observableProto.bufferTime = bufferTime_1.bufferTime;
var bufferToggle_1 = require('./operators/bufferToggle');
observableProto.bufferToggle = bufferToggle_1.bufferToggle;
var bufferWhen_1 = require('./operators/bufferWhen');
observableProto.bufferWhen = bufferWhen_1.bufferWhen;
var catch_1 = require('./operators/catch');
observableProto.catch = catch_1._catch;
var combineAll_1 = require('./operators/combineAll');
observableProto.combineAll = combineAll_1.combineAll;
var combineLatest_1 = require('./operators/combineLatest');
observableProto.combineLatest = combineLatest_1.combineLatest;
var concat_1 = require('./operators/concat');
observableProto.concat = concat_1.concat;
var concatAll_1 = require('./operators/concatAll');
observableProto.concatAll = concatAll_1.concatAll;
var concatMap_1 = require('./operators/concatMap');
observableProto.concatMap = concatMap_1.concatMap;
var concatMapTo_1 = require('./operators/concatMapTo');
observableProto.concatMapTo = concatMapTo_1.concatMapTo;
var count_1 = require('./operators/count');
observableProto.count = count_1.count;
var dematerialize_1 = require('./operators/dematerialize');
observableProto.dematerialize = dematerialize_1.dematerialize;
var debounce_1 = require('./operators/debounce');
observableProto.debounce = debounce_1.debounce;
var debounceTime_1 = require('./operators/debounceTime');
observableProto.debounceTime = debounceTime_1.debounceTime;
var defaultIfEmpty_1 = require('./operators/defaultIfEmpty');
observableProto.defaultIfEmpty = defaultIfEmpty_1.defaultIfEmpty;
var delay_1 = require('./operators/delay');
observableProto.delay = delay_1.delay;
var distinctUntilChanged_1 = require('./operators/distinctUntilChanged');
observableProto.distinctUntilChanged = distinctUntilChanged_1.distinctUntilChanged;
var do_1 = require('./operators/do');
observableProto.do = do_1._do;
var expand_1 = require('./operators/expand');
observableProto.expand = expand_1.expand;
var filter_1 = require('./operators/filter');
observableProto.filter = filter_1.filter;
var finally_1 = require('./operators/finally');
observableProto.finally = finally_1._finally;
var first_1 = require('./operators/first');
observableProto.first = first_1.first;
var groupBy_1 = require('./operators/groupBy');
observableProto.groupBy = groupBy_1.groupBy;
var ignoreElements_1 = require('./operators/ignoreElements');
observableProto.ignoreElements = ignoreElements_1.ignoreElements;
var every_1 = require('./operators/every');
observableProto.every = every_1.every;
var last_1 = require('./operators/last');
observableProto.last = last_1.last;
var map_1 = require('./operators/map');
observableProto.map = map_1.map;
var mapTo_1 = require('./operators/mapTo');
observableProto.mapTo = mapTo_1.mapTo;
var materialize_1 = require('./operators/materialize');
observableProto.materialize = materialize_1.materialize;
var merge_1 = require('./operators/merge');
observableProto.merge = merge_1.merge;
var mergeAll_1 = require('./operators/mergeAll');
observableProto.mergeAll = mergeAll_1.mergeAll;
var mergeMap_1 = require('./operators/mergeMap');
observableProto.mergeMap = mergeMap_1.mergeMap;
observableProto.flatMap = mergeMap_1.mergeMap;
var mergeMapTo_1 = require('./operators/mergeMapTo');
observableProto.mergeMapTo = mergeMapTo_1.mergeMapTo;
observableProto.flatMapTo = mergeMapTo_1.mergeMapTo;
var multicast_1 = require('./operators/multicast');
observableProto.multicast = multicast_1.multicast;
var observeOn_1 = require('./operators/observeOn');
observableProto.observeOn = observeOn_1.observeOn;
var partition_1 = require('./operators/partition');
observableProto.partition = partition_1.partition;
var publish_1 = require('./operators/publish');
observableProto.publish = publish_1.publish;
var publishBehavior_1 = require('./operators/publishBehavior');
observableProto.publishBehavior = publishBehavior_1.publishBehavior;
var publishReplay_1 = require('./operators/publishReplay');
observableProto.publishReplay = publishReplay_1.publishReplay;
var reduce_1 = require('./operators/reduce');
observableProto.reduce = reduce_1.reduce;
var repeat_1 = require('./operators/repeat');
observableProto.repeat = repeat_1.repeat;
var retry_1 = require('./operators/retry');
observableProto.retry = retry_1.retry;
var retryWhen_1 = require('./operators/retryWhen');
observableProto.retryWhen = retryWhen_1.retryWhen;
var sample_1 = require('./operators/sample');
observableProto.sample = sample_1.sample;
var sampleTime_1 = require('./operators/sampleTime');
observableProto.sampleTime = sampleTime_1.sampleTime;
var scan_1 = require('./operators/scan');
observableProto.scan = scan_1.scan;
var share_1 = require('./operators/share');
observableProto.share = share_1.share;
var single_1 = require('./operators/single');
observableProto.single = single_1.single;
var skip_1 = require('./operators/skip');
observableProto.skip = skip_1.skip;
var skipUntil_1 = require('./operators/skipUntil');
observableProto.skipUntil = skipUntil_1.skipUntil;
var startWith_1 = require('./operators/startWith');
observableProto.startWith = startWith_1.startWith;
var subscribeOn_1 = require('./operators/subscribeOn');
observableProto.subscribeOn = subscribeOn_1.subscribeOn;
var switch_1 = require('./operators/switch');
observableProto.switch = switch_1._switch;
var switchMap_1 = require('./operators/switchMap');
observableProto.switchMap = switchMap_1.switchMap;
var switchMapTo_1 = require('./operators/switchMapTo');
observableProto.switchMapTo = switchMapTo_1.switchMapTo;
var take_1 = require('./operators/take');
observableProto.take = take_1.take;
var takeUntil_1 = require('./operators/takeUntil');
observableProto.takeUntil = takeUntil_1.takeUntil;
var throttle_1 = require('./operators/throttle');
observableProto.throttle = throttle_1.throttle;
var timeout_1 = require('./operators/timeout');
observableProto.timeout = timeout_1.timeout;
var timeoutWith_1 = require('./operators/timeoutWith');
observableProto.timeoutWith = timeoutWith_1.timeoutWith;
var toArray_1 = require('./operators/toArray');
observableProto.toArray = toArray_1.toArray;
var toPromise_1 = require('./operators/toPromise');
observableProto.toPromise = toPromise_1.toPromise;
var window_1 = require('./operators/window');
observableProto.window = window_1.window;
var windowCount_1 = require('./operators/windowCount');
observableProto.windowCount = windowCount_1.windowCount;
var windowTime_1 = require('./operators/windowTime');
observableProto.windowTime = windowTime_1.windowTime;
var windowToggle_1 = require('./operators/windowToggle');
observableProto.windowToggle = windowToggle_1.windowToggle;
var windowWhen_1 = require('./operators/windowWhen');
observableProto.windowWhen = windowWhen_1.windowWhen;
var withLatestFrom_1 = require('./operators/withLatestFrom');
observableProto.withLatestFrom = withLatestFrom_1.withLatestFrom;
var zip_1 = require('./operators/zip');
observableProto.zip = zip_1.zipProto;
var zipAll_1 = require('./operators/zipAll');
observableProto.zipAll = zipAll_1.zipAll;
/* tslint:disable:no-unused-variable */
var Subject_1 = require('./Subject');
exports.Subject = Subject_1.Subject;
var Subscription_1 = require('./Subscription');
exports.Subscription = Subscription_1.Subscription;
var Subscriber_1 = require('./Subscriber');
exports.Subscriber = Subscriber_1.Subscriber;
var ReplaySubject_1 = require('./subjects/ReplaySubject');
exports.ReplaySubject = ReplaySubject_1.ReplaySubject;
var BehaviorSubject_1 = require('./subjects/BehaviorSubject');
exports.BehaviorSubject = BehaviorSubject_1.BehaviorSubject;
var ConnectableObservable_1 = require('./observables/ConnectableObservable');
exports.ConnectableObservable = ConnectableObservable_1.ConnectableObservable;
var Notification_1 = require('./Notification');
exports.Notification = Notification_1.Notification;
var EmptyError_1 = require('./util/EmptyError');
exports.EmptyError = EmptyError_1.EmptyError;
var ArgumentOutOfRangeError_1 = require('./util/ArgumentOutOfRangeError');
exports.ArgumentOutOfRangeError = ArgumentOutOfRangeError_1.ArgumentOutOfRangeError;
var nextTick_1 = require('./schedulers/nextTick');
var immediate_1 = require('./schedulers/immediate');
/* tslint:enable:no-unused-variable */
/* tslint:disable:no-var-keyword */
var Scheduler = {
    nextTick: nextTick_1.nextTick,
    immediate: immediate_1.immediate
};
exports.Scheduler = Scheduler;
/* tslint:enable:no-var-keyword */
